BUG: FreeBSD fails a conversion test

FreeBSD recently began failing the `test_nested_arrays_stringlength` test in `test_array_coercion.py`. It emits a `RuntimeWarning` when casting `float` objects to `string` dtype

```
self = <test_array_coercion.TestStringDiscovery object at 0x1a16d2f95dd0>
obj = 1.2

    @pytest.mark.parametrize("obj",
            [object(), 1.2, 10**43, None, "string"],
            ids=["object", "1.2", "10**43", "None", "string"])
    def test_nested_arrays_stringlength(self, obj):
        length = len(str(obj))
        expected = np.dtype(f"S{length}")
        arr = np.array(obj, dtype="O")
>       assert np.array([arr, arr], dtype="S").dtype == expected
E       RuntimeWarning: invalid value encountered in cast

arr        = array(1.2, dtype=object)
expected   = dtype('S3')
length     = 3
obj        = 1.2
self       = <test_array_coercion.TestStringDiscovery object at 0x1a16d2f95dd0>

../.venv/lib/python3.11/site-packages/numpy/_core/tests/test_array_coercion.py:180: RuntimeWarning
```

Maybe we could get some FreeBSD eyeballs on this, I don't know who we should ping. I see @oscargus  and @DavidAlphaFox have previously contributed FreeBSD issues and PRs, maybe they know who to turn to? In any case, #28331 makes CI pass by checking that this warning is emitted. This issue is to remind us to try to fix it.

Maybe this is a NumPy bug after all? Or I've done something strange.

I tried creating a standalone example, see: https://github.com/ngoldbaum/cpython_bug

You should be able to build and run it with:

```
meson setup build
cd build
ninja
./cpython_bug
```

And if everything is correct it should print without exiting.

Interestingly, when I try building on FreeBSD I see:

```
1/1] Linking target cpython_bug
FAILED: cpython_bug 
cc  -o cpython_bug cpython_bug.p/main.c.o -Wl,--as-needed -Wl,--no-undefined -Wl,--start-group /root/numpy/.venv/lib/python3.11/site-packages/numpy/_core/include/../lib/libnpymath.a -L/usr/local/lib -lpython3.11 -Wl,--end-group
ld: error: undefined symbol: feraiseexcept
>>> referenced by ieee754.c.src:389 (../numpy/_core/src/npymath/ieee754.c.src:389)
>>>               meson-generated_ieee754.c.o:(npy_set_floatstatus_divbyzero) in archive /root/numpy/.venv/lib/python3.11/site-packages/numpy/_core/include/../lib/libnpymath.a
>>> referenced by ieee754.c.src:394 (../numpy/_core/src/npymath/ieee754.c.src:394)
>>>               meson-generated_ieee754.c.o:(npy_set_floatstatus_overflow) in archive /root/numpy/.venv/lib/python3.11/site-packages/numpy/_core/include/../lib/libnpymath.a
>>> referenced by ieee754.c.src:399 (../numpy/_core/src/npymath/ieee754.c.src:399)
>>>               meson-generated_ieee754.c.o:(npy_set_floatstatus_underflow) in archive /root/numpy/.venv/lib/python3.11/site-packages/numpy/_core/include/../lib/libnpymath.a
>>> referenced 1 more times
cc: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

Indicating that somehow the C++ version of this file is getting linked in? It looks like NumPy distributes both an `ieee754.c.src` and `ieee754.cpp` file - are somehow the state of the two modules getting mixed up in a way that causes linker errors here but incorrect results inside NumPy, which includes symbols for both versions?

Just speculation though.

--- a/numpy/_core/src/multiarray/convert_datatype.c
+++ b/numpy/_core/src/multiarray/convert_datatype.c
@@ -3507,7 +3507,9 @@ initialize_void_and_object_globals(void) {
     method->nin = 1;
     method->nout = 1;
     method->name = "object_to_any_cast";
-    method->flags = NPY_METH_SUPPORTS_UNALIGNED | NPY_METH_REQUIRES_PYAPI;
+    method->flags = (NPY_METH_SUPPORTS_UNALIGNED
+                     | NPY_METH_REQUIRES_PYAPI
+                     | NPY_METH_NO_FLOATINGPOINT_ERRORS);
     method->casting = NPY_UNSAFE_CASTING;
     method->resolve_descriptors = &object_to_any_resolve_descriptors;
     method->get_strided_loop = &object_to_any_get_loop;
@@ -3522,7 +3524,9 @@ initialize_void_and_object_globals(void) {
     method->nin = 1;
     method->nout = 1;
     method->name = "any_to_object_cast";
-    method->flags = NPY_METH_SUPPORTS_UNALIGNED | NPY_METH_REQUIRES_PYAPI;
+    method->flags = (NPY_METH_SUPPORTS_UNALIGNED
+                     | NPY_METH_REQUIRES_PYAPI
+                     | NPY_METH_NO_FLOATINGPOINT_ERRORS);
     method->casting = NPY_SAFE_CASTING;
     method->resolve_descriptors = &any_to_object_resolve_descriptors;
     method->get_strided_loop = &any_to_object_get_loop;

--- a/numpy/_core/src/multiarray/dtype_transfer.c
+++ b/numpy/_core/src/multiarray/dtype_transfer.c
@@ -236,7 +236,7 @@ any_to_object_get_loop(
         NPY_ARRAYMETHOD_FLAGS *flags)
 {
 
-    *flags = NPY_METH_REQUIRES_PYAPI;  /* No need for floating point errors */
+    *flags = NPY_METH_REQUIRES_PYAPI | NPY_METH_NO_FLOATINGPOINT_ERRORS;
 
     *out_loop = _strided_to_strided_any_to_object;
     *out_transferdata = PyMem_Malloc(sizeof(_any_to_object_auxdata));
@@ -342,7 +342,7 @@ object_to_any_get_loop(
         NpyAuxData **out_transferdata,
         NPY_ARRAYMETHOD_FLAGS *flags)
 {
-    *flags = NPY_METH_REQUIRES_PYAPI;
+    *flags = NPY_METH_REQUIRES_PYAPI | NPY_METH_NO_FLOATINGPOINT_ERRORS;
 
     /* NOTE: auxdata is only really necessary to flag `move_references` */
     _object_to_any_auxdata *data = PyMem_Malloc(sizeof(*data));

--- a/numpy/_core/src/multiarray/dtype_transfer.c
+++ b/numpy/_core/src/multiarray/dtype_transfer.c
@@ -235,7 +235,7 @@ any_to_object_get_loop(
         NpyAuxData **out_transferdata,
         NPY_ARRAYMETHOD_FLAGS *flags)
 {
-
+    /* Python API doesn't use FPEs and this also attempts to hide spurious ones. */
     *flags = NPY_METH_REQUIRES_PYAPI | NPY_METH_NO_FLOATINGPOINT_ERRORS;
 
     *out_loop = _strided_to_strided_any_to_object;
@@ -342,6 +342,7 @@ object_to_any_get_loop(
         NpyAuxData **out_transferdata,
         NPY_ARRAYMETHOD_FLAGS *flags)
 {
+    /* Python API doesn't use FPEs and this also attempts to hide spurious ones. */
     *flags = NPY_METH_REQUIRES_PYAPI | NPY_METH_NO_FLOATINGPOINT_ERRORS;
 
     /* NOTE: auxdata is only really necessary to flag `move_references` */

